src/pudl/metadata/classes.py

```diff
@@ -0,0 +1,911 @@
+"""Metadata data classes."""
+import copy
+import datetime
+import uuid
+from typing import (Any, Callable, Dict, Iterable, List, Literal, Optional,
+                    Tuple, Type, Union)
+
+import pandas as pd
+import pydantic
+import sqlalchemy as sa
+
+from .constants import (CONTRIBUTORS, CONTRIBUTORS_BY_SOURCE, FIELD_DTYPES,
+                        FIELD_DTYPES_SQL, KEYWORDS_BY_SOURCE, LICENSES,
+                        PERIODS, SOURCES)
+from .fields import FIELDS
+from .helpers import (expand_periodic_column_names, groupby_aggregate,
+                      has_duplicate_basenames, most_and_more_frequent,
+                      split_period)
+from .resources import FOREIGN_KEYS, RESOURCES
+
+# ---- Base ---- #
+
+
+class Base(pydantic.BaseModel):
+    """
+    Custom Pydantic base class.
+
+    It overrides :meth:`fields` and :meth:`schema` to allow properties with those names.
+    To use them in a class, use an underscore prefix and an alias.
+
+    Examples:
+        >>> class Class(Base):
+        ...     fields_: List[str] = pydantic.Field(alias="fields")
+        >>> m = Class(fields=['x'])
+        >>> m
+        Class(fields=['x'])
+        >>> m.fields
+        ['x']
+        >>> m.fields = ['y']
+        >>> m.dict()
+        {'fields': ['y']}
+    """
+
+    class Config:
+        """Custom Pydantic configuration."""
+
+        validate_all: bool = True
+        validate_assignment: bool = True
+        extra: str = 'forbid'
+
+    def dict(self, *args, by_alias=True, **kwargs) -> dict:  # noqa: A003
+        """Return as a dictionary."""
+        return super().dict(*args, by_alias=by_alias, **kwargs)
+
+    def json(self, *args, by_alias=True, **kwargs) -> str:
+        """Return as JSON."""
+        return super().json(*args, by_alias=by_alias, **kwargs)
+
+    def __getattribute__(self, name: str) -> Any:
+        """Get attribute."""
+        if name in ("fields", "schema") and f"{name}_" in self.__dict__:
+            name = f"{name}_"
+        return super().__getattribute__(name)
+
+    def __setattr__(self, name, value) -> None:
+        """Set attribute."""
+        if name in ("fields", "schema") and f"{name}_" in self.__dict__:
+            name = f"{name}_"
```

Comment:
Yes, this is part of the hacky workaround needed to allow Pydantic models to have attributes with the name `fields` or `schema` (which are otherwise methods of `pydantic.BaseModel`).